src/library/scala/collection/immutable/HashMap.scala

```diff
@@ -209,7 +210,17 @@ sealed class HashMap[A, +B] extends AbstractMap[A, B]
         // so we invert
         (this.merge0(thatHash, 0, merger.invert)).asInstanceOf[That]
 
-        // cope with Maps that have a foreachEntry ??
+      case that:HasForeachEntry[A, B] =>
+        object adder extends Function2[A, B, Unit] {
+          var result: HashMap[A, B] = this.asInstanceOf[HashMap[A, B]]
+          val merger = HashMap.liftMerger[A, B](null)
+
+          override def apply(key: A, value: B): Unit = {
+            result = result.updated0(key, computeHash(key), 0, value, null, merger)
```

Comment:
Is this really better than using `foreach`? Now you don't get a tuple anymore (even when the original map has one) so you have to compute a new one later on demand.

Reply:
Its a trade off. maybe the foreachEntry should have a tuple if it has one?
It depends if you generate the tuple, and how many Maps actually have the tuple I think

Reply:
the only classes that implement `HasForEachEntry` are 
`HashMap` - covered previously in matches
`ListMap` - doesn't  maintain tuples
`SortedMap`/`RedBlackTree` - doesn't maintain tuples 

which collections types are you concerned with, when we have the existing tuple and lose it?